### PR TITLE
Morebits: more documentation

### PR DESCRIPTION
**Pull request by siddharthvp**
Continuation of #481. JSDocs for queryString, quickForm and simpleWindow.

There are still a handful of undocumented functions remaining though.

### CODE REVIEW

File: morebits.js [hunk 1]
```diff
@@ -770,24 +781,43 @@ Morebits.quickForm.setElementLabel = function QuickFormSetElementLabel(element,
...
 	if (element.hasAttribute("data-oldlabel")) {
 		return Morebits.quickForm.setElementLabel(element, element.getAttribute("data-oldlabel"));
 	}
 	return null;
 };
 
+/**
+ * Shows or hides a form element plus its label and tooltip
+ * @param {*} element
+ * @param {*} visibility
```
Comment (MusikAnimal) [inline morebits.js]: I'm guessing `element` would have to either be of type `string` or `jQuery`? `visibility` I assume would be `boolean`?

File: morebits.js [hunk 2]
```diff
@@ -2413,9 +2467,9 @@ Morebits.wiki.page = function(pageName, currentAction) {
 	};
 
 	/* Private member functions
-	 *
-	 * These are not exposed outside
-	 */
+		*
+		* These are not exposed outside
+		*/
```
Comment (MusikAnimal) [inline morebits.js]: Spacing

File: morebits.js [hunk 3]
```diff
@@ -2943,11 +2997,11 @@ Morebits.wiki.page = function(pageName, currentAction) {
...
- * - Should we retry loads also?
- * - Need to reset current action before the save?
- * - Deal with action.completed stuff
- * - Need to reset all parameters once done (e.g. edit summary, move destination, etc.)
- */
+	* - Should we retry loads also?
+	* - Need to reset current action before the save?
+	* - Deal with action.completed stuff
+	* - Need to reset all parameters once done (e.g. edit summary, move destination, etc.)
+	*/
```
Comment (MusikAnimal) [inline morebits.js]: Spacing

File: morebits.js [hunk 4]
```diff
@@ -2961,10 +3015,10 @@ Morebits.wiki.page = function(pageName, currentAction) {
...
  */
 
- /**
-  * @constructor
-  * @param {HTMLDivElement} previewbox - the <div> element that will contain the rendered HTML
-  */
+	/**
+	 * @constructor
+	 * @param {HTMLDivElement} previewbox - the <div> element that will contain the rendered HTML
+	 */
```
Comment (MusikAnimal) [inline morebits.js]: Spacing

File: morebits.js [hunk 5]
```diff
@@ -3237,29 +3293,22 @@ Morebits.wikitext.page.prototype = {
...
  *
  * var value = Morebits.queryString.get('key');
  * var obj = new Morebits.queryString('foo=bar&baz=quux');
  * value = obj.get('foo');
  */
+
+	/**
+	 * @constructor
+	 * @param {string} qString   The query string
+	 */
```
Comment (MusikAnimal) [inline morebits.js]: Spacing

File: morebits.js [hunk 6]
```diff
@@ -3355,6 +3388,40 @@ Morebits.queryString.create = function( arr ) {
...
+Morebits.queryString.prototype.get = function(key) {
+	return this.params[key] ? this.params[key] : null;
+};
+
+/**
+ * @returns true if the particular `key` is set
+ * @param {string} key
+ */
+Morebits.queryString.prototype.exists = function(key) {
+	return this.params[key] ? true : false;
```
Comment (MusikAnimal) [inline morebits.js]: How about `!!this.params[key]`?

File: morebits.js [hunk 7]
```diff
@@ -3355,6 +3388,40 @@ Morebits.queryString.create = function( arr ) {
...
+};
+
+/**
+ * @returns true if the value associated with given `key` equals given `value`
+ * @param {string} key
+ * @param {string} value
+ */
+Morebits.queryString.prototype.equals = function(key, value) {
+	return this.params[key] === value ? true : false;
```
Comment (MusikAnimal) [inline morebits.js]: I don't think we need the ternary operator, the `===` will return true/false

File: morebits.js [hunk 8]
```diff
@@ -3355,6 +3388,40 @@ Morebits.queryString.create = function( arr ) {
 	}
 	return resarr.join('&');
 };
+
+/**
+ * @returns the value associated to the given `key`
+ * @param {string} key
+ */
+Morebits.queryString.prototype.get = function(key) {
+	return this.params[key] ? this.params[key] : null;
```
Comment (MusikAnimal) [inline morebits.js]: Could do `this.params[key] || null` but the above is probably more clear

File: morebits.js [hunk 9]
```diff
@@ -3634,14 +3707,30 @@ Morebits.batchOperation = function(currentAction) {
...
 	this.setPageList = function(pageList) {
 		ctx.pageList = pageList;
 	};
 
+	/**
+	 * Sets a known option:
+	 * - chunkSize (integer): 	the size of chunks to break the array into (default 50).
+	 * 							Setting this to a small value (<5) can cause problems.
+	 * - preserveIndividualStatusLines (boolean): keep each page's status element visible
+	 *                                            when worker is complete?
```
Comment (MusikAnimal) [inline morebits.js]: Some weird spacing here

File: morebits.js [hunk 10]
```diff
@@ -3849,7 +3948,11 @@ Morebits.simpleWindow.prototype = {
 
 		return this;
 	},
-	// Shows the dialog.  Calling display() on a dialog that has previously been closed might work, but it is not guaranteed.
+
+	/**
+	 * Shows the dialog. Calling display() on a dialog that has previously been closed
+	 * might work, but it is not guaranteed.
+	 */
```
Comment (MusikAnimal) [inline morebits.js]: `@returns {Morebits}` (I think)? Same for the others below

Comment (siddharthvp) [reply morebits.js]: `@returns {Morebits.simpleWinkdow}`

File: morebits.js (hunk 1 above)
Comment (siddharthvp) [reply morebits.js]: On testing, element can be HTMLElement or jQuery, visibility: boolean

Comment (MusikAnimal) [reply morebits.js]: The code here would seem to allow for a string (the CSS selector). Just pointing this out; if only HTMLElement and jQuery makes sense here that's totally fine :)

Comment (siddharthvp) [reply morebits.js]: oh yeah, added it

**Comment by siddharthvp** [comment]
@MusikAnimal   Sorry for all those excess spacings, dunno how they came about. I hope everything is fine now. With the last commit, I think we've got _almost_ the whole of the Morebits library covered. The only stuff w/o proper documentation now is `Morebits.status` and `Morebits.bytes`, plus a few functions here and there, like `Morebits.wikitext.template.parse` and `Morebits.quickForm.element.prototype.compute`.

**Pull request merged**